sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncClient.java

```diff
@@ -411,4 +421,101 @@ public CosmosAsyncDatabase getDatabase(String id) {
+                Optional<Object> callerFunc = reactorContext.getOrEmpty(TracerProvider.NESTED_CALL);
+                if (!callerFunc.isPresent()) {
+                    parentContext.set(tracerProvider.startSpan(spanName,
+                        context.addData(TracerProvider.ATTRIBUTE_MAP, tracingAttributes), ProcessKind.DATABASE));
+                }
+            }
+        }).doOnSuccess(response -> {
+            if (isTracingEnabled) {
+                tracerProvider.endSpan(parentContext.get(), Signal.complete(), response.getStatusCode());
+            }
+        }).doOnError(throwable -> {
+            if (isTracingEnabled) {
+                tracerProvider.endSpan(parentContext.get(), Signal.error(throwable), 0);
+            }
+        });
+    }
+
+    private Mono<CosmosAsyncDatabaseResponse> createDatabase(Database database, CosmosDatabaseRequestOptions options, Context context){
+        final boolean isTracingEnabled = tracerProvider.isEnabled();
+        final AtomicReference<Context> parentContext = isTracingEnabled
+            ? new AtomicReference<>(Context.NONE)
+            : null;
+        String spanName = "createDatabase." + database.getId();
+        Map<String, String> tracingAttributes = new HashMap<String, String>() {{
+            put(TracerProvider.DB_TYPE, TracerProvider.DB_TYPE_VALUE);
+            put(TracerProvider.DB_INSTANCE, database.getId());
+            put(TracerProvider.DB_URL, serviceEndpoint);
+            put(TracerProvider.DB_STATEMENT, spanName);
+        }};
+
+        return asyncDocumentClient.createDatabase(database, ModelBridgeInternal.toRequestOptions(options))
+            .map(databaseResourceResponse -> ModelBridgeInternal.createCosmosAsyncDatabaseResponse(databaseResourceResponse,
+                this))
+            .single()
+            .doOnSubscribe(ignoredValue -> {
+                if (isTracingEnabled) {
+                    reactor.util.context.Context reactorContext = FluxUtil.toReactorContext(context);
+                    Objects.requireNonNull(reactorContext.hasKey(TracerProvider.MASTER_CALL));
+                    Optional<Object> callerFunc = reactorContext.getOrEmpty(TracerProvider.NESTED_CALL);
+                    if (!callerFunc.isPresent()) {
+                        parentContext.set(tracerProvider.startSpan(spanName,
+                            context.addData(TracerProvider.ATTRIBUTE_MAP, tracingAttributes), ProcessKind.DATABASE));
+                    }
+                }
+            }).doOnSuccess(response -> {
+                if (isTracingEnabled) {
+                    tracerProvider.endSpan(parentContext.get(), Signal.complete(), response.getStatusCode());
+                }
+            }).doOnError(throwable -> {
+                if (isTracingEnabled) {
+                    tracerProvider.endSpan(parentContext.get(), Signal.error(throwable), 0);
+                }
+            });
+    }
```

Comment:
@simplynaveen20 - It seems the trace specific `doOnSubscribe | doOnSuccess | doOnError `  handlers need to be repeated in all APIs. Wondering it makes sense to centralize it

With that this method looks something like:

```java
private Mono<CosmosAsyncTriggerResponse> read(Context context) {
    Mono<CosmosAsyncDatabaseResponse> resultPublisher = asyncDocumentClient.createDatabase(database, ModelBridgeInternal.toRequestOptions(options))
        .map(databaseResourceResponse -> ModelBridgeInternal.createCosmosAsyncDatabaseResponse(databaseResourceResponse,
            this))
        .single()
    if (!tracerProvider.isEnabled()) {
        return resultPublisher;
    } else {
        final String spanName = "createDatabase." + database.getId();
        Map<String, String> tracingAttributes = new HashMap<String, String>() {{
            put(TracerProvider.DB_TYPE, TracerProvider.DB_TYPE_VALUE);
            put(TracerProvider.DB_INSTANCE, database.getId());
            put(TracerProvider.DB_URL, serviceEndpoint);
            put(TracerProvider.DB_STATEMENT, spanName);
        }};
    return traceEnabledPublisher(resultPublisher, tracingAttributes);
    }
}

// Reusable code 
//
private <T> Mono<T> traceEnabledPublisher(Mono<T> resultPublisher, Map<String, String> tracingAttributes) {
    return resultPublisher
        .doOnSubscribe(ignoredValue -> {
                reactor.util.context.Context reactorContext = FluxUtil.toReactorContext(context);
                Objects.requireNonNull(reactorContext.hasKey(TracerProvider.MASTER_CALL));
                Optional<Object> callerFunc = reactorContext.getOrEmpty(TracerProvider.NESTED_CALL);
                if (!callerFunc.isPresent()) {
                    parentContext.set(tracerProvider.startSpan(spanName,
                        context.addData(TracerProvider.ATTRIBUTE_MAP, tracingAttributes), ProcessKind.DATABASE));
                }
        }).doOnSuccess(response -> {
            tracerProvider.endSpan(parentContext.get(), Signal.complete(), response.getStatusCode());
        }).doOnError(throwable -> {
            tracerProvider.endSpan(parentContext.get(), Signal.error(throwable), 0);
        });
}

```

Reply:
Yea actually I am thinking of cleaning this , and some other ones too  .Thank for suggestion . Will send it in next iteration

Reply:
Done